src/rust/ide/lib/view/src/graph_editor/component/breadcrumbs.rs

```diff
@@ -0,0 +1,229 @@
+}
+
+impl FrpInputs {
+    /// Create new FrpInputs.
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            def push_breadcrumb = source();
+            def pop_breadcrumb  = source();
+        }
+        Self{push_breadcrumb,pop_breadcrumb}
+    }
+}
+
+
+
+// ==================
+// === FrpOutputs ===
+// ==================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct FrpOutputs {
+    pub pushed_breadcrumb : frp::Source<String>,
+    pub popped_breadcrumb : frp::Source<String>
+}
+
+impl FrpOutputs {
+    /// Create new FrpOutputs.
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            def pushed_breadcrumb = source();
+            def popped_breadcrumb = source();
+        }
+        Self{pushed_breadcrumb,popped_breadcrumb}
+    }
+}
+
+
+
+// ===========
+// === Frp ===
+// ===========
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct Frp {
+    pub inputs  : FrpInputs,
+    pub outputs : FrpOutputs,
+    pub network : frp::Network,
+}
+
+impl Deref for Frp {
+    type Target = FrpInputs;
+    fn deref(&self) -> &Self::Target {
+        &self.inputs
+    }
+}
+
+impl Default for Frp {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Frp {
+    /// Create new Frp.
+    pub fn new() -> Self {
+        let network = frp::Network::new();
+        let inputs  = FrpInputs::new(&network);
+        let outputs = FrpOutputs::new(&network);
+        Self{network,inputs,outputs}
+    }
+}
+
+
+
+// ==================
+// === Animations ===
+// ==================
+
+/// ProjectName's animations handlers.
+#[derive(Debug,Clone,CloneRef,Copy)]
+pub struct Animations {}
+
+impl Animations {
+    /// Create new animations handlers.
+    pub fn new(_network:&frp::Network) -> Self {
+        Self{}
+    }
+}
+
+
+
+// ========================
+// === BreadcrumbsModel ===
+// ========================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct BreadcrumbsModel {
+    logger         : Logger,
+    animations     : Animations,
+    display_object : display::object::Instance,
+    scene          : Scene,
+    breadcrumbs    : Rc<RefCell<Vec<Breadcrumb>>>
+}
+
+impl BreadcrumbsModel {
+    /// Create new ProjectNameModel.
+    pub fn new<'t,S:Into<&'t Scene>>(scene:S,frp:&Frp) -> Self {
+        let scene          = scene.into();
+        let logger         = Logger::new("Breadcrumbs");
+        let display_object = display::object::Instance::new(&logger);
+        let animations     = Animations::new(&frp.network);
+        let scene          = scene.clone_ref();
+        let breadcrumbs    = Rc::new(RefCell::new(default()));
+        Self{logger,display_object,animations,scene,breadcrumbs}
+    }
+
+    fn width(&self) -> f32 {
+        let mut width = 0.0;
+        for breadcrumb in self.breadcrumbs.borrow().iter() {
+            width += breadcrumb.width();
+        }
+        width
```

Comment:
not sure if this isnt overcomplicating, but you could use iterator's `map(...).sum()` or some fold.